Prevent deadlocking due to errors

Oozlum opened the pull request:
Prevent errors in protocol.read_response from propagating before the
current request has been removed from the fifo.  Otherwise all other
pipelined requests will become deadlocked.

Oozlum committed:
Prevent deadlocking due to errors

Prevent errors in protocol.read_response from propagating before the
current request has been removed from the fifo.  Otherwise all other
pipelined requests will become deadlocked.

daurnimator reviewed (commented):
I think using a pcall here breaks under certain environments. I had to do a big workaround in lua-http to make pcall work in Lua 5.1:

https://github.com/daurnimator/lua-http/blob/54606a9d3fea9780468601ed0c8be48748e97204/http/util.lua#L210

Is there some other way we can solve this without a `pcall`?

Oozlum committed:
Refactor command sending.

protocol.lua now handles all low-level socket read/write operations,
rather than being split across two files.

Oozlum committed:
Use custom CQueues socket:onerror handler.

Ensure all socket errors are returned to caller using nil, error
convention, rather than directly calling assert() or error()

Oozlum committed:
Make encode_request, send_command and read_response non-asserting

These functions now use the 'return nil, error' convention for reporting
errors.  This allows us to avoid using pcall for error handling, which
doesn't play well with CQueues on PUC Lua 5.1 in certain situations.

Oozlum commented:
> I think using a pcall here breaks under certain environments. I had to do a big workaround in lua-http to make pcall work in Lua 5.1:
> 
> https://github.com/daurnimator/lua-http/blob/54606a9d3fea9780468601ed0c8be48748e97204/http/util.lua#L210
> 
> Is there some other way we can solve this without a `pcall`?

Yes, actually, but the changes are a bit more invasive.  I've committed a different implementation which uses a custom cqueues.socket:onerror handler and changes the internal error mechanism to use the 'return nil, error' convention rather than being assert() based.  This allows handling of errors without using pcall.

daurnimator commented on lredis/protocol.lua:
```diff
@@ -11,7 +11,10 @@ end
 -- Requests are always just an array of bulk strings
 local function encode_request(arg)
 	local n = arg.n or #arg
-	assert(n > 0, "need at least one argument")
```
invalid input should remain as `asserts`: only runtime/environment inflicted errors should `return nil, err`

Oozlum replied:
We **cannot** assert here.  If we do, then there is no choice but to wrap the calling code in a ```pcall()``` to catch it, and that is exactly what we're trying to avoid because of the Lua 5.1 issues.  The error is `error()`'d back to the caller at a higher level, after handling the error in a Lua 5.1 compatible way.

daurnimator commented on lredis/cqueues.lua:
```diff
@@ -11,7 +11,14 @@ local mt = {
+-- override the default socket handler so that it returns all errors rather than
+-- throwing them.
+local function socket_error_handler(socket, method, code, level)
+	return code, level
```
might need more than this, e.g.
  - turning it into a string
  - prettier TLS errors
  - handling timeouts

Oozlum replied:
The existing handler does not do any of these things.  This is the same interface used by the ```cqueues``` default handler.  The only difference is that this does ```return code, level``` for all errors; ```cqueues``` only does this for ```ETIMEDOUT``` and ```EPIPE``` but ```assert()```'s everything else, which again is what we're trying to avoid.

daurnimator commented on lredis/cqueues.lua:
```diff
@@ -87,7 +100,10 @@ function methods:get_next(new_status, new_error, string_null, array_null)
 	end
-	local resp = protocol.read_response(self.socket, new_status, new_error, string_null, array_null)
+	local null = {}
+	local resp, err_code = protocol.read_response(self.socket, new_status, new_error, null, array_null)
```
This changes important semantics: why not pass `string_null`?

Oozlum replied:
Because ```protocol.read_response``` now does ```return nil, error``` on error and if ```string_null``` happens to be ```nil``` (which is perfectly valid) then we have no way to differentiate the two.  We need to pass a new temporary ```null``` so that we can do that.

There is no change to semantics here.  If you look further down, after checking for errors there is a comparison to ```null``` and we revert to ```string_null``` in that case.

daurnimator commented on lredis/protocol.lua:
```diff
@@ -36,43 +39,81 @@ local function encode_inline(arg)
-	assert(line:sub(-1, -1) == "\r", "invalid line ending")
+	local line, err_code = file:read("*l")
+	if not line then
+		return nil, err_code or "EOF reached"
```
Sometimes this is a number, sometimes a string; need to be consistent on error types

Oozlum replied:
If ```file:read()``` fails then the default ```cqueues``` handler may ```assert()``` with a numeric code.  If ```file:read()``` returns something unexpected then your code ```assert()```'s a string.  There is no further inconsistency introduced by this change.

daurnimator commented on lredis/protocol.lua:
```diff
@@ -36,43 +39,81 @@ local function encode_inline(arg)
 		return new_error(data)
 	elseif status == ":" then
-		return assert(tonumber(data, 10), "invalid integer")
+		return tonumber(data, 10), "invalid integer"
```
don't return an error message when an error didn't occur

Oozlum replied:
Ok, happy to change this.

daurnimator commented on lredis/protocol.lua:
```diff
@@ -36,43 +39,81 @@ local function encode_inline(arg)
 			return string_null
 		elseif len > 512*1024*1024 then -- max 512 MB
-			error("bulk string too large")
+			return nil, "bulk string too large"
```
may need to invent error codes?

Oozlum replied:
Why?  In what way is this any different to the original code?  After error handling takes place safely at a higher level the result is `error()`'d back to the caller.  There is no change to what the caller receives as a result of this change.

daurnimator commented on lredis/protocol.lua:
```diff
@@ -36,43 +39,81 @@ local function encode_inline(arg)
 			return array_null
 		else
-			local arr = {}
+			local arr, null = {}, {}
```
don't invent your own `string_null`

Oozlum replied:
Again, we need it to differentiate from ```string_null == nil```.  See my previous response.

Oozlum committed:
Don't return error string if there is no error.
